// sigma/renderers/webgl/programs/edge.fast.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Sigma.js WebGL Renderer Fast Edge Program
 * ==========================================
 *
 * Program rendering edges using GL_LINES which is presumably very fast but
 * won't render thickness correctly on some GPUs and has some quirks.
 */
var program_1 = __importDefault(require("./program"));
var utils_1 = require("../utils");
var edge_fast_vert_glsl_1 = __importDefault(require("../shaders/edge.fast.vert.glsl.js"));
var edge_fast_frag_glsl_1 = __importDefault(require("../shaders/edge.fast.frag.glsl.js"));
var POINTS = 2, ATTRIBUTES = 3;
var EdgeFastProgram = /** @class */ (function (_super) {
    __extends(EdgeFastProgram, _super);
    function EdgeFastProgram(gl) {
        var _this = _super.call(this, gl, edge_fast_vert_glsl_1.default, edge_fast_frag_glsl_1.default) || this;
        // Locations
        _this.positionLocation = gl.getAttribLocation(_this.program, "a_position");
        _this.colorLocation = gl.getAttribLocation(_this.program, "a_color");
        _this.resolutionLocation = gl.getUniformLocation(_this.program, "u_resolution");
        _this.matrixLocation = gl.getUniformLocation(_this.program, "u_matrix");
        // Bindings
        gl.enableVertexAttribArray(_this.positionLocation);
        gl.enableVertexAttribArray(_this.colorLocation);
        gl.vertexAttribPointer(_this.positionLocation, 2, gl.FLOAT, false, ATTRIBUTES * Float32Array.BYTES_PER_ELEMENT, 0);
        gl.vertexAttribPointer(_this.colorLocation, 1, gl.FLOAT, false, ATTRIBUTES * Float32Array.BYTES_PER_ELEMENT, 8);
        return _this;
    }
    EdgeFastProgram.prototype.allocate = function (capacity) {
        this.array = new Float32Array(POINTS * ATTRIBUTES * capacity);
    };
    EdgeFastProgram.prototype.process = function (sourceData, targetData, data, offset) {
        var array = this.array;
        var i = 0;
        if (sourceData.hidden || targetData.hidden || data.hidden) {
            for (var l = i + POINTS * ATTRIBUTES; i < l; i++)
                array[i] = 0;
        }
        var x1 = sourceData.x, y1 = sourceData.y, x2 = targetData.x, y2 = targetData.y, color = utils_1.floatColor(data.color);
        i = POINTS * ATTRIBUTES * offset;
        // First point
        array[i++] = x1;
        array[i++] = y1;
        array[i++] = color;
        // Second point
        array[i++] = x2;
        array[i++] = y2;
        array[i] = color;
    };
    EdgeFastProgram.prototype.bufferData = function () {
        var gl = this.gl;
        // Vertices data
        gl.bufferData(gl.ARRAY_BUFFER, this.array, gl.DYNAMIC_DRAW);
    };
    EdgeFastProgram.prototype.render = function (params) {
        var gl = this.gl;
        var program = this.program;
        gl.useProgram(program);
        // Binding uniforms
        gl.uniform2f(this.resolutionLocation, params.width, params.height);
        gl.uniformMatrix3fv(this.matrixLocation, false, params.matrix);
        // Drawing:
        gl.drawArrays(gl.LINES, 0, this.array.length / ATTRIBUTES);
    };
    return EdgeFastProgram;
}(program_1.default));
exports.default = EdgeFastProgram;
