// sigma/renderers/webgl/programs/node.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Sigma.js WebGL Renderer Node Program
 * =====================================
 *
 * Simple program rendering nodes as discs, shaped by triangles using the
 * `gl.TRIANGLES` display mode. So, to draw one node, it will need to store
 * three times the center of the node, with the color, the size and an angle
 * indicating which "corner" of the triangle to draw.
 */
var program_1 = __importDefault(require("./program"));
var utils_1 = require("../utils");
var node_vert_glsl_1 = __importDefault(require("../shaders/node.vert.glsl.js"));
var node_frag_glsl_1 = __importDefault(require("../shaders/node.frag.glsl.js"));
var ANGLE_1 = 0, ANGLE_2 = (2 * Math.PI) / 3, ANGLE_3 = (4 * Math.PI) / 3;
var POINTS = 3, ATTRIBUTES = 5;
var NodeProgram = /** @class */ (function (_super) {
    __extends(NodeProgram, _super);
    function NodeProgram(gl) {
        var _this = _super.call(this, gl, node_vert_glsl_1.default, node_frag_glsl_1.default) || this;
        // Locations
        _this.positionLocation = gl.getAttribLocation(_this.program, "a_position");
        _this.sizeLocation = gl.getAttribLocation(_this.program, "a_size");
        _this.colorLocation = gl.getAttribLocation(_this.program, "a_color");
        _this.angleLocation = gl.getAttribLocation(_this.program, "a_angle");
        _this.resolutionLocation = gl.getUniformLocation(_this.program, "u_resolution");
        _this.matrixLocation = gl.getUniformLocation(_this.program, "u_matrix");
        _this.ratioLocation = gl.getUniformLocation(_this.program, "u_ratio");
        _this.scaleLocation = gl.getUniformLocation(_this.program, "u_scale");
        // Bindings
        gl.enableVertexAttribArray(_this.positionLocation);
        gl.enableVertexAttribArray(_this.sizeLocation);
        gl.enableVertexAttribArray(_this.colorLocation);
        gl.enableVertexAttribArray(_this.angleLocation);
        gl.vertexAttribPointer(_this.positionLocation, 2, gl.FLOAT, false, ATTRIBUTES * Float32Array.BYTES_PER_ELEMENT, 0);
        gl.vertexAttribPointer(_this.sizeLocation, 1, gl.FLOAT, false, ATTRIBUTES * Float32Array.BYTES_PER_ELEMENT, 8);
        gl.vertexAttribPointer(_this.colorLocation, 1, gl.FLOAT, false, ATTRIBUTES * Float32Array.BYTES_PER_ELEMENT, 12);
        gl.vertexAttribPointer(_this.angleLocation, 1, gl.FLOAT, false, ATTRIBUTES * Float32Array.BYTES_PER_ELEMENT, 16);
        return _this;
    }
    NodeProgram.prototype.allocate = function (capacity) {
        this.array = new Float32Array(POINTS * ATTRIBUTES * capacity);
    };
    NodeProgram.prototype.process = function (data, offset) {
        var color = utils_1.floatColor(data.color);
        var i = offset * POINTS * ATTRIBUTES;
        var array = this.array;
        array[i++] = data.x;
        array[i++] = data.y;
        array[i++] = data.size;
        array[i++] = color;
        array[i++] = ANGLE_1;
        array[i++] = data.x;
        array[i++] = data.y;
        array[i++] = data.size;
        array[i++] = color;
        array[i++] = ANGLE_2;
        array[i++] = data.x;
        array[i++] = data.y;
        array[i++] = data.size;
        array[i++] = color;
        array[i] = ANGLE_3;
    };
    NodeProgram.prototype.bufferData = function () {
        var gl = this.gl;
        gl.bufferData(gl.ARRAY_BUFFER, this.array, gl.DYNAMIC_DRAW);
    };
    NodeProgram.prototype.render = function (params) {
        var gl = this.gl;
        var program = this.program;
        gl.useProgram(program);
        gl.uniform2f(this.resolutionLocation, params.width, params.height);
        gl.uniform1f(this.ratioLocation, 1 / Math.pow(params.ratio, params.nodesPowRatio));
        gl.uniform1f(this.scaleLocation, params.scalingRatio);
        gl.uniformMatrix3fv(this.matrixLocation, false, params.matrix);
        gl.drawArrays(gl.TRIANGLES, 0, this.array.length / ATTRIBUTES);
    };
    return NodeProgram;
}(program_1.default));
exports.default = NodeProgram;
